atc/db/resource_check_rate_limiter.go

```diff
@@ -45,15 +42,16 @@ func NewResourceCheckRateLimiter(
 		limiter.checkInterval = checkInterval
 		limiter.refreshConn = refreshConn
 		limiter.refreshLimiter = rate.NewLimiter(rate.Every(refreshInterval), 1)
+
+		// The first time we call Wait, we will properly update the limit.
+		// This is just to avoid dealing with the limiter not existing.
+		limiter.checkLimiter = rate.NewLimiter(rate.Inf, 1)
 	}
 
 	return limiter
 }
 
 func (limiter *ResourceCheckRateLimiter) Wait(ctx context.Context) error {
-	limiter.mut.Lock()
-	defer limiter.mut.Unlock()
-
```

Comment:
While the only race condition I can think of is the `limter.checkLimiter.SetLimit()` getting called twice with different values, I still think it would be better keep this mutex just for sanity. If we get rid of the defer and explicitly call Unlock on line 58 and 63, it should solve the inconsistent delay.

Reply:
I guess the implications of guarding the refresh with a mutex is that later calls to `Wait` will have to wait for the limit to be updated before proceeding, which seems reasonable. Updated